mongo/src/main/java/com/redhat/lightblue/mongo/crud/Utils.java

```diff
@@ -0,0 +1,71 @@
+/*
+ Copyright 2013 Red Hat, Inc. and/or its affiliates.
+
+ This file is part of lightblue.
+
+ This program is free software: you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ This program is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package com.redhat.lightblue.mongo.crud;
+
+import java.util.Set;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.ArrayList;
+
+import com.mongodb.DBCollection;
+import com.mongodb.DBCursor;
+import com.mongodb.BasicDBObject;
+
+public final class Utils {
+
+    /**
+     * Returns the _ids of documents that are failed to be updated
+     *
+     * @param collection The DB collection
+     * @param currentTxId The txid of the current operation. The
+     * docversion field of all docs are set to this value for the
+     * current operation
+     * @param idToTxIdMap A mapping of doc _id to the old transaction
+     * id of the document
+     *
+     * @return A set of _ids whose `document version is not equal to currentTxId
+     */
+    public static Set<Object> checkFailedUpdates(DBCollection collection,
+                                                 String currentTxId,
+                                                 Map<Object,String> idToTxIdMap) {
+        // We would like to find all _id:currentTxId documents. Those
+        // that are missing, i.e. not in the resultset, are the docs
+        // that are modified during or after our update operation
+        DBCursor cursor=null;
+        try {
+            Set<Object> keys=idToTxIdMap.keySet();
+            cursor=collection.find(new BasicDBObject(Translator.DOC_VERSION_FULLPATH_STR,currentTxId).
```

Comment:
It'd return partial already. The problem is the updates that did not fail,
but reported as such, because they might force a retry even though the
update worked.

On Thu, Aug 11, 2016 at 4:06 PM, Alec Henninger <notifications@github.com>
wrote:

> In mongo/src/main/java/com/redhat/lightblue/mongo/crud/Utils.java
> <https://github.com/lightblue-platform/lightblue-mongo/pull/299#discussion_r74512766>
> :
>
> > +     * current operation
> > +     * @param idToTxIdMap A mapping of doc _id to the old transaction
> > +     * id of the document
> > +     *
> > +     * @return A set of _ids whose `document version is not equal to currentTxId
> > +     */
> > +    public static Set<Object> checkFailedUpdates(DBCollection collection,
> > +                                                 String currentTxId,
> > +                                                 Map<Object,String> idToTxIdMap) {
> > +        // We would like to find all _id:currentTxId documents. Those
> > +        // that are missing, i.e. not in the resultset, are the docs
> > +        // that are modified during or after our update operation
> > +        DBCursor cursor=null;
> > +        try {
> > +            Set<Object> keys=idToTxIdMap.keySet();
> > +            cursor=collection.find(new BasicDBObject(Translator.DOC_VERSION_FULLPATH_STR,currentTxId).
>
> What if for a request where concurrent mod is detected, we return with
> status "partial"? It is still preferable to make the concurrent mod
> detection visible to client over silently replace the effects of a previous
> successful update IMHO, even if the client doesn't get to see what records
> did not get the update. Additionally, I'd like to be able to get some
> reporting on how often this case actually happens.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/lightblue-platform/lightblue-mongo/pull/299/files/cd27c2bf02d26eeb2d30652940c19a45acc212f1#r74512766>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADgDDcYJg8U9NkhKaH5qSzdukSMp03Ozks5qe5zegaJpZM4JfOdd>
> .
>